File: csharp/src/HeapSort.cs

using System.Collections.Generic;

static class MaxHeapSort<T> where T : IComparable<T>
{
    static public void Execute(List<T> arr)
    {
        for (int idx = (arr.Count / 2) - 1; idx >= 0; --idx)
        {
            HeapSort(arr, arr.Count, idx);
        }

        for (int idx = arr.Count - 1; idx > 0; --idx)
        {
            (arr[0], arr[idx]) = (arr[idx], arr[0]);
            HeapSort(arr, idx, 0);
        }
    }

    static private void HeapSort(List<T> arr, int cnt, int idx)
    {
        int largest = idx;
        int left = (2 * idx) + 1;
        int right = (2 * idx) + 2;

        if (left < cnt && arr[largest].CompareTo(arr[left]) < 0)
        {
            largest = left;
        }

        if (right < cnt && arr[largest].CompareTo(arr[right]) < 0)
        {
            largest = right;
        }

        if (largest != idx)
        {
            (arr[largest], arr[idx]) = (arr[idx], arr[largest]);
            HeapSort(arr, cnt, largest);
        }
    }
}

static class MinHeapSort<T> where T : IComparable<T>
{
    static public void Execute(List<T> arr)
    {
        for (int idx = (arr.Count / 2) - 1; idx >= 0; --idx)
        {
            HeapSort(arr, arr.Count, idx);
        }

        for (int idx = arr.Count - 1; idx > 0; --idx)
        {
            (arr[0], arr[idx]) = (arr[idx], arr[0]);
            HeapSort(arr, idx, 0);
        }
    }

    static private void HeapSort(List<T> arr, int cnt, int idx)
    {
        int largest = idx;
        int left = (2 * idx) + 1;
        int right = (2 * idx) + 2;

        if (left < cnt && arr[largest].CompareTo(arr[left]) > 0)
        {
            largest = left;
        }

        if (right < cnt && arr[largest].CompareTo(arr[right]) > 0)
        {
            largest = right;
        }

        if (largest != idx)
        {
            (arr[largest], arr[idx]) = (arr[idx], arr[largest]);
            HeapSort(arr, cnt, largest);
        }
    }
}

File: go/src/sort/heapsort.go
package sort

import (
	"cmp"
)

func MaxHeapSort[T cmp.Ordered](arr []T) {
	n := len(arr)

	// NOTE. 논리적 트리 생성
	for i := n/2 - 1; i >= 0; i-- {
		heapifyMax(arr, n, i)
	}

	// NOTE. 정렬
	for i := n - 1; i > 0; i-- {
		// NOTE. 맨 앞 데이터는 맨 뒤로 Shift
		arr[0], arr[i] = arr[i], arr[0]
		// NOTE. Shift 데이터를 제외한 나머지 데이터로 논리적 트리 생성
		heapifyMax(arr, i, 0)
	}
}

func heapifyMax[T cmp.Ordered](arr []T, n, i int) {
	largest := i
	left := (2 * i) + 1
	right := (2 * i) + 2

	if left < n && arr[left] > arr[largest] {
		largest = left
	}

	if right < n && arr[right] > arr[largest] {
		largest = right
	}

	if largest != i {
		arr[i], arr[largest] = arr[largest], arr[i]
		heapifyMax(arr, n, largest)
	}
}

func MinHeapSort(arr []int) {
	n := len(arr)
	for i := n/2 - 1; i >= 0; i-- {
		heapifyMin(arr, n, i)
	}
	for i := n - 1; i > 0; i-- {
		arr[0], arr[i] = arr[i], arr[0]
		heapifyMin(arr, i, 0)
	}
}

func heapifyMin(arr []int, n, i int) {
	smallest := i
	left := (2 * i) + 1
	right := (2 * i) + 2

	if left < n && arr[left] < arr[smallest] {
		smallest = left
	}

	if right < n && arr[right] < arr[smallest] {
		smallest = right
	}

	if smallest != i {
		arr[i], arr[smallest] = arr[smallest], arr[i]
		heapifyMin(arr, n, smallest)
	}
}
